src/cppsim/circuit_optimizer.cpp

```diff
@@ -198,3 +235,680 @@ QuantumGateMatrix* QuantumCircuitOptimizer::merge_all(
+    // bool is_matched(UINT logical_qubit)
+
+    UINT num_inserted_gates = 0;
+    int j = lowest_idx - 1;
+    for (int i = local_qc - 1; i >= (int)lowest_idx; --i) {
+        // if the qubit meets condition, skip it
+        if (is_matched(qt.p2l[i])) continue;
+
+        // if the qubit does not meet condition, find a matched qubit
+        for (; j >= 0; --j) {
+            if (is_matched(qt.p2l[j])) break;
+        }
+
+        // swap qubits if a matched qubit is found
+        if (j >= 0) {
+            num_inserted_gates += qt.add_swap_gate(
+                circuit, j, i, 1, gate_insertion_pos + num_inserted_gates);
+        } else {
+            throw std::runtime_error(
+                "Error: "
+                "QuantumCircuitOptimizer::move_matching_qubits_to_local_upper()"
+                " "
+                "no enougth matched qubits");
+        }
+    }
+    return num_inserted_gates;
+}
+
+UINT QuantumCircuitOptimizer::rearrange_qubits(const UINT gate_idx,
+    std::unordered_set<UINT> next_local_qubit, QubitTable& qt) {
+    UINT num_inserted_gates = 0;
+
+    std::unordered_set<UINT> cur_local_qubits(
+        qt.p2l.cbegin(), qt.p2l.cbegin() + local_qc);
+    std::unordered_set<UINT> import_qubits, exportable_qubits;
+
+    // e.g.
+    // qt.p2l = [0 4 3 1 2 7 6 5]
+    // global_qc = 4, local_qc = 4
+    //
+    //  local  | global
+    // -----------------
+    // 0 4 3 1 | 2 7 6 5
+    //
+    // cur_local_qubits = {0, 1, 3 ,4}
+    // next_local_qubits = {2, 4, 6}
+
+    // import_qubits = next_local_qubit - cur_local_qubits
+    // e.g. {2, 6} = {2, 4, 6} - {0, 1, 3 ,4}
+    for (UINT i : next_local_qubit) {
+        if (cur_local_qubits.count(i) == 0) {
+            import_qubits.insert(i);
+        }
+    }
+
+    // exportable_qubits = cur_local_qubits - next_local_qubit
+    // e.g. {0, 1, 3} = {0, 1, 3 ,4} - {2, 4, 6}
+    for (UINT i : cur_local_qubits) {
+        if (next_local_qubit.count(i) == 0) {
+            exportable_qubits.insert(i);
+        }
+    }
+
+    LOG << "import qubits: " << to_string(import_qubits) << std::endl;
+    LOG << "exportable qubits: " << to_string(exportable_qubits) << std::endl;
+
+    // make index and width for swap
+    // e.g. swap_global_idx_list = [2 6]
+    //      swap_width_list      = [1 1]
+    std::vector<UINT> swap_global_idx_list;
+    std::vector<UINT> swap_width_list;
+    for (UINT i = local_qc; i < circuit->qubit_count; i++) {
+        if (import_qubits.count(qt.p2l[i])) {
+            swap_global_idx_list.push_back(i);
+            swap_width_list.push_back(1);
+        }
+    }
+
+    // fuse swaps
+    // e.g. swap_global_idx_list = [2 6] -> [2]
+    //      swap_width_list      = [1 1] -> [3]
+    // TODO: if fast qubit swapping between global qubits is implemented,
+    //       global qubits should be rearranged to be contiguous.
+    UINT total_swap_width = import_qubits.size();
+    int extra_local_qc = exportable_qubits.size() - import_qubits.size();
```

Comment:
All supported compilers express signed integers as Two's Complement so this works, but it is forced from C++20.
`(int)exportable_qubits.size() - (int)import_qubits.size()` is safer.